File: InterviewPreparationKit/src/string/manipulation/SpecialStringAgain.java
package string.manipulation;

public class SpecialStringAgain {

    public static void main(String[] args) {
        long result = substrCount2(2, "aaa");
        System.out.println("result: " + result);
    }

    static long substrCount(int n, String s) {
        long specialSubStringNum = 0;


        return specialSubStringNum;
    }

    public static long substrCount2(int length, String s) {
        long counter = 0;
        for (int i = 0; i < length; i++) {
            // if the current symbol is in the middle of palindrome, e.g. aba
            int offset = 1;
            while (i - offset >= 0
                    && i + offset < length
                    && s.charAt(i - offset) == s.charAt(i - 1)
                    && s.charAt(i + offset) == s.charAt(i - 1)) {

                counter++;
                offset++;
            }
            // if this is repeatable characters aa
            int repeats = 0;
            while (i + 1 < length && s.charAt(i) == s.charAt(i + 1)) {
                repeats++;
                i++;
            }
            counter += repeats * (repeats + 1) / 2;
        }
        return counter + length;
    }
}

File: InterviewPreparationKit/src/string/manipulation/SpecialStringAgain.kt
package string.manipulation

fun main() {
    val str = "mnonopoo"
    print(substrCount(str.length, str))
}

fun substrCount(length: Int, s: String): Long {
    var counter: Long = 0
    var i = 0
    while (i < length) {

        // if the current symbol is in the middle of palindrome, e.g. aba
        var offset = 1
        while (i - offset >= 0 && i + offset < length && s[i - offset] == s[i - 1] && s[i + offset] == s[i - 1]) {
            counter++
            offset++
        }
        // if this is repeatable characters aa
        var repeats = 0
        while (i + 1 < length && s[i] == s[i + 1]) {
            repeats++
            i++
        }
        counter += (repeats * (repeats + 1) / 2).toLong()
        i++
    }
    return counter + length
}